transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -802,22 +827,86 @@ private void write(Object msg, boolean flush, ChannelPromise promise) {
+                            // release the message.
+                            promise.setFailure(e);
+                        } finally {
+                            ReferenceCountUtil.release(m);
+                        }
+                        return;
+                    }
+                }
+                // We scheduled a flush so now one is pending.
+                next.invokeWrite(m, promise, true);
             }
         } else {
-            AbstractWriteTask task;
-            if (flush) {
-                task = WriteAndFlushTask.newInstance(next, m, promise);
-            }  else {
-                task = WriteTask.newInstance(next, m, promise);
+            next.invokeWrite(m, promise, false);
+        }
+    }
+
+    private void writeFromOutsideEventLoop(AbstractChannelHandlerContext next, EventExecutor executor, Object m,
+                                           boolean flush, ChannelPromise promise) {
+        if (flush) {
+            safeExecute(executor, WriteAndFlushTask.newInstance(next, m, promise, false), promise, m);
+        } else if (isAutoFlush()) {
+            if (!(executor() instanceof SingleThreadEventLoop)) {
+                // The Executor is not of type SingleThreadEventLoop which means we have no way to schedule a
+                // flush at the end of the event loop run. Just explicitly do a flush in this case.
+                safeExecute(executor, WriteAndFlushTask.newInstance(next, m, promise, true), promise, m);
+            } else {
+                // We first add the write task and then schedule something on the loop if needed. This is done
+                // so we are sure we not miss the flush. We will explicit handle the case of
+                // RejectedExecutionException in this case.
+                safeExecute(executor, WriteTask.newInstance(next, m, promise, true), promise, m);
+
+                // Only schedule if there is no other flush pending yet.
+                if (!flushPending) {
+                    try {
+                        scheduleFlushOnEventLoopIteration();
+                    } catch (RejectedExecutionException e) {
+                        // If it was rejected but the promise is already completed we not care too much, otherwise
+                        // release the message.
+                        if (promise.tryFailure(e)) {
+                            ReferenceCountUtil.release(m);
+                        }
+                    }
+                }
             }
-            safeExecute(executor, task, promise, m);
+        } else {
+            safeExecute(executor, WriteTask.newInstance(next, m, promise, false), promise, m);
         }
     }
 
+    private boolean isAutoFlush() {
+        // Check for null to not produce a NPE if the ChannelConfig implementation not supports the ChannelOption.
+        Boolean autoFlush = channel().config().getOption(ChannelOption.AUTO_FLUSH);
+        return autoFlush != null ? autoFlush : false;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Ternary operators that can be simplified with || or && [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)